Python/Product/VSCode/AnalysisVsc/LanguageServer.Lifetime.cs

```diff
@@ -0,0 +1,134 @@
+
+        [JsonRpcMethod("shutdown")]
+        public Task Shutdown() => _server.Shutdown();
+
+        [JsonRpcMethod("exit")]
+        public async Task Exit() {
+            await _server.Exit();
+            _sessionTokenSource.Cancel();
+        }
+
+        private Task LoadDirectoryFiles(InitializeParams @params) {
+            string rootDir = null;
+            if (@params.rootUri != null) {
+                rootDir = @params.rootUri.ToAbsolutePath();
+            } else if (!string.IsNullOrEmpty(@params.rootPath)) {
+                rootDir = PathUtils.NormalizePath(@params.rootPath);
+            }
+
+            if (string.IsNullOrEmpty(rootDir)) {
+                return Task.CompletedTask;
+            }
+
+            var matcher = new Matcher();
+            matcher.AddIncludePatterns(@params.initializationOptions.includeFiles ?? new[] { "**/*" });
+            matcher.AddExcludePatterns(@params.initializationOptions.excludeFiles ?? Enumerable.Empty<string>());
+
+            var dib = new DirectoryInfoWrapper(new DirectoryInfo(rootDir));
+            var matchResult = matcher.Execute(dib);
+
+            _server.LogMessage(MessageType.Log, $"Loading files from {rootDir}");
+            return LoadFromDirectoryAsync(rootDir, matchResult);
+        }
+
+        private async Task LoadFromDirectoryAsync(string rootDir, PatternMatchingResult matchResult) {
+            foreach (var file in matchResult.Files) {
+                if (_sessionTokenSource.IsCancellationRequested) {
+                    break;
+                }
+
+                var path = Path.Combine(rootDir, file.Path);
+                if (!ModulePath.IsPythonSourceFile(path)) {
+                    if (ModulePath.IsPythonFile(path, true, true, true)) {
+                        // TODO: Deal with scrapable files (if we need to do anything?)
+                    }
+                    continue;
+                }
+                await _server.LoadFileAsync(new Uri(path));
+            }
+        }
+
+        private void MonitorParentProcess(InitializeParams p) {
+            // Monitor parent process
+            Process parentProcess = null;
+            if (p.processId.HasValue) {
+                try {
+                    parentProcess = Process.GetProcessById(p.processId.Value);
+                } catch (ArgumentException) { }
+
+                Debug.Assert(parentProcess != null, "Parent process does not exist");
+                if (parentProcess != null) {
+                    parentProcess.Exited += (s, e) => {
+                        _sessionTokenSource.Cancel();
+                    };
+                }
+            }
+
+            if (parentProcess != null) {
+                Task.Run(async () => {
+                    while (!_sessionTokenSource.IsCancellationRequested) {
+                        await Task.Delay(2000);
+                        if (parentProcess.HasExited) {
+                            _sessionTokenSource.Cancel();
```

Comment:
This whole task can probably be simplified to `parentProcess.Exited += (s, e) => _sessionTokenSource.Cancel()`

Reply:
I see you already had it there... so why the loop as well?

Reply:
On Mac parentProcess.Exited never happens.